### ISSUE TITLE
🔍 PCA Methods Reliability Audit for v1.1.0

### ISSUE THREAD

**Issue by bitjungle** [bitjungle/gopca#443]
# 🔍 PCA Methods Reliability Audit for v1.1.0

## Overview
Before releasing v1.1.0, we must ensure 100% reliability across all PCA implementations (SVD, NIPALS, Kernel, Temporal). This audit will validate mathematical correctness, numerical stability, and consistency across methods.

## Motivation
GoPCA aims to be the definitive PCA toolkit—reliable, general-purpose, and accessible. Each method must be mathematically correct, numerically stable, and produce results consistent with established reference implementations.

#### A. Reference Implementation Comparison
Each method must match reference implementations within acceptable tolerance:

- [ ] **Standard PCA (SVD/NIPALS)**
  - Reference: R `prcomp()`, Python `sklearn.decomposition.PCA`
  - Tolerance: 1e-10 for well-conditioned matrices
  - Test: Iris, Wine, synthetic datasets

- [ ] **Kernel PCA**
  - Reference: Python `sklearn.decomposition.KernelPCA`
  - Test kernels: RBF, polynomial, sigmoid
  - Test: Swiss Roll, concentric circles, moons dataset

- [ ] **Temporal PCA (SSA)**
  - Reference: R `Rssa` package
  - Test: CO2 data, sunspot data, synthetic AR/MA series
  - Verify lag matrix construction matches Golyandina et al. (2018)

#### B. Algorithm Properties Verification
- [ ] Orthogonality of components: `|U'U - I| < 1e-12`
- [ ] Variance explained sums to ≤100%
- [ ] Eigenvalues in descending order
- [ ] Reconstruction error decreases with more components

#### A. Edge Cases
- [ ] Single sample (n=1)
- [ ] Single variable (p=1)
- [ ] n < p (more variables than samples)
- [ ] p < n (more samples than variables)
- [ ] Perfect correlation (rank-deficient)
- [ ] Near-zero variance variables
- [ ] Missing values (where applicable)

#### B. Conditioning Tests
- [ ] Well-conditioned matrix (condition number < 100)
- [ ] Moderately ill-conditioned (100 < κ < 10^6)
- [ ] Severely ill-conditioned (κ > 10^6)
- [ ] Compare SVD vs eigendecomposition stability

#### C. Scale Sensitivity
- [ ] Variables with vastly different scales (10^-6 to 10^6)
- [ ] Negative values
- [ ] Zero-centered vs positive-only data
- [ ] Log-scale vs linear-scale data

#### A. Preprocessing Pipeline
Verify for each method that preprocessing is applied at the correct stage:

```
Standard PCA: Data → Preprocess → SVD/NIPALS
Kernel PCA:   Data → Preprocess → Kernel → Eigendecomposition
Temporal PCA: Data → Preprocess → Lag Matrix → SVD
```

- [ ] Mean-centering
- [ ] Standardization (z-score)
- [ ] Robust scaling (median/MAD)
- [ ] Scale-only (no centering)
- [ ] SNV (row-wise standardization)

#### B. Transform Consistency
- [ ] `fit()` then `transform()` gives same result as `fit_transform()`
- [ ] Transform on training data reproduces training scores
- [ ] Transform on new data maintains preprocessing parameters

### 4. Cross-Method Validation 🔀

When applicable, different methods should give similar results:

- [ ] SVD vs NIPALS on standard data (should match exactly)
- [ ] Standard PCA vs Kernel PCA with linear kernel
- [ ] Temporal PCA with lag=1 vs standard PCA (conceptually different but worth documenting)

### 5. Performance Benchmarks ⚡

Document performance characteristics:

- [ ] Time complexity verification
  - SVD: O(min(n²p, np²))
  - NIPALS: O(npc) where c is iterations
  - Kernel: O(n³)
  - Temporal: O((n-L+1) × (pL)²)

- [ ] Memory usage
  - Peak memory for 10K×1K matrix
  - Memory scaling with components

### 6. Real-World Dataset Validation 🌍

Test with datasets from different domains:

#### Standard Datasets
- [ ] **Iris** (n=150, p=4): Classic, well-conditioned
- [ ] **Wine** (n=178, p=13): Mixed scales
- [ ] **Corn** (n=80, p=700): Spectroscopy, p>>n

#### Domain-Specific
- [ ] **Financial**: Stock prices (multiplicative, needs log)
- [ ] **Meteorological**: MET data (mixed units, temporal)
- [ ] **Chemical Process**: Autoclave sensors (outliers, batch)
- [ ] **Spectroscopy**: NIR/UV-Vis (high-dimensional)

### 7. Documentation Audit 📚

- [ ] Each method references authoritative sources:
  - SVD/NIPALS: Jolliffe & Cadima (2016), Bro & Smilde (2014)
  - Kernel PCA: Schölkopf et al. (1998)
  - Temporal PCA: Ghil et al. (2002), Golyandina et al. (2018)

- [ ] Preprocessing effects documented
- [ ] When to use which method
- [ ] Limitations and assumptions

### 8. Error Handling & User Guidance 🛡️

- [ ] Clear error messages for invalid input
- [ ] Warnings for numerical instability
- [ ] Suggestions for preprocessing when scales differ >100x
- [ ] Automatic method recommendation based on data characteristics

### Phase 1: Unit Tests
Create comprehensive test suite:
```go
// temporal_pca_validation_test.go
func TestTemporalPCAAgainstRssa(t *testing.T) {
    // Load reference results from Rssa
    // Compare with tolerance
}
```

### Phase 2: Integration Tests
Full pipeline validation with real datasets

### Phase 3: Benchmark Suite
Performance and accuracy benchmarks

## Success Criteria

✅ All mathematical properties verified
✅ Numerical stability within documented limits
✅ Matches reference implementations (within tolerance)
✅ Performance meets or exceeds targets
✅ Documentation complete and accurate
✅ 100% test coverage for core algorithms

## References

Available in `docs_tmp/references/`:
- Jolliffe & Cadima (2016) - Principal component analysis: a review
- Shlens (2014) - A Tutorial on Principal Component Analysis
- Esbensen et al. (2002) - Multivariate Data Analysis in Practice
- Bro & Smilde (2014) - Principal component analysis
- Ghil et al. (2002) - Advanced spectral methods for climatic time series
- Golyandina et al. - Multivariate and 2D Extensions of SSA with Rssa

## Timeline

- Week 1-2: Implement test suite
- Week 3: Run validation, document findings
- Week 4: Fix issues, re-validate
- Week 5: Documentation updates
- Week 6: Final review and sign-off

## Assigned

- [ ] Core team review
- [ ] Assign validation lead
- [ ] External review (optional but recommended)

---

**Priority**: 🔴 Critical for v1.1.0
**Labels**: `validation`, `testing`, `reliability`, `v1.1.0-blocker`

**Comment by bitjungle** [comment]
## ✅ Phase 3 Complete: Cross-method validation tests

Phase 3 has been successfully completed and merged to develop via PR #479.

### Implemented Tests
- **SVD vs NIPALS consistency**: Validates that both methods produce equivalent results for various condition numbers
- **Preprocessing consistency**: Ensures all PCA methods apply preprocessing identically  
- **Temporal PCA validation**: Tests SSA-based temporal decomposition
- **Component selection consistency**: Verifies automatic component selection based on variance explained
- **Kernel PCA framework**: Test framework ready (implementation needs fixes for true linear kernel equivalence)

### Test Coverage
All tests compile and pass successfully:
- ✅ TestSVDvsNIPALS - PASS (κ=10, 100, 1000, 10000)
- ✅ TestPreprocessingConsistency - PASS
- ✅ TestTemporalPCAVsStandard - PASS
- ✅ TestComponentSelectionConsistency - PASS
- ⚠️ TestKernelPCALinearVsStandard - Framework ready, needs kernel implementation fixes

### Progress Summary
- [x] Phase 1: Core reliability validation tests (#465) ✅
- [x] Phase 2: Stability tests and numerical behavior (#466) ✅  
- [x] **Phase 3: Cross-method validation tests (#467)** ✅
- [ ] Phase 4: Transform and preprocessing validation (#468)
- [ ] Phase 5: Edge cases and error handling tests (#469)

The reliability audit is progressing well with 3 of 5 phases now complete.

**Comment by bitjungle** [comment]
## Cross-Method Validation Tests Completed ✅

PR #482 has been created to implement cross-method validation tests (issue #467), which is part of this reliability audit.

### What was implemented:

**Cross-Method Validation Tests:**
- SVD vs NIPALS comparison for different condition numbers
- Preprocessing consistency across methods
- Temporal PCA validation against standard PCA
- Component selection consistency tests

**Data Consistency Tests:**
- Method consistency across iris and wine datasets
- Preprocessing reversibility verification
- Transform operation consistency
- Missing value handling validation

**Test Infrastructure:**
- Reusable comparison functions with sign ambiguity handling
- Orthogonality and unit norm verification utilities
- Condition-number aware tolerance calculations

### Key Points:
- All tests pass (83.9% coverage maintained)
- Follows Core Development Principles (DRY, existing patterns)
- Uses existing  function - no code duplication
- Proper mathematical validation with appropriate tolerances

This completes the cross-method validation portion of the reliability audit as outlined in issue #467.

Related PRs:
- #481 - Reverted broken implementation
- #482 - Proper implementation of cross-method validation tests

**Comment by bitjungle** [comment]
All four phases of the PCA reliability audit are complete:

- Phase 1 (#465): Core reliability validation
- Phase 2 (#466): Stability and numerical behavior
- Phase 3 (#467): Cross-method validation (SVD vs NIPALS)
- Phase 4 (#480): Transform, preprocessing pipeline, and model persistence (PR #625)

`internal/core/` test coverage: 88.3%. Both transform paths (in-memory and CLI JSON persistence) are validated across all preprocessing types. Ready to proceed with structural refactoring (#459).
